### PR TITLE
Delete redundant controllers and endpoints; add new endpoints etc.

### PR DESCRIPTION
**Pull request by Alexy-Def**

### CODE REVIEW

**Review by wollentimer — APPROVED** [review]

File: BLL/DTO/CandidateDTO.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@ public class CandidateDTO
         public string Skype { get; set; }
         public int SpecializationID { get; set; }
         public int CityID { get; set; }
+        public int CountryID { get; set; }
```
Comment (AnnKirienko) [inline BLL/DTO/CandidateDTO.cs:21]: You changed candidate dto model and candidate model but you didn't change methods in endpoints. Please, delete this changes, if you are not going to do it

File: BLL/Services/DirectoryService.cs [hunk 2]
```diff
@@ -32,15 +33,27 @@ public List<SpecializationDTO> GetAllSpecializations()
             return _mapper.Map<List<Specialization>, List<SpecializationDTO>>(_db.Specializations.GetAll().ToList());
         }
 
+        public void SaveSpecialization(List<SpecializationDTO> specializationsDto)
+        {
+            _db.Specializations.RemoveAll();
```
Comment (VasiliNesterenko) [inline BLL/Services/DirectoryService.cs:38]: it's not a good idea to remove all records
does any other table have references to Specialization? if yes, you cannot remove records from Specialization until you drop all references in the other tables

File: BLL/Services/DirectoryService.cs [hunk 3]
```diff
@@ -88,19 +101,41 @@ public CityDTO GetCityById(int cityId)
             return _mapper.Map<City, CityDTO>(_db.Cities.Get(cityId));
         }
 
+        public string GetLocationById(int cityId)
+        {
+            var city = _db.Cities.Get(cityId);
```
Comment (VasiliNesterenko) [inline BLL/Services/DirectoryService.cs:106]: no need to make 3 calls to the DB, you can do it with only one:

var city = _db.Cities.FindWithSpecificationPattern(new CitySpecification()).FirstOrDefault(x => x.Id == cityId);
return $"{city.State.Country.CountryName}, {city.State.StateName}, {city.CityName}";

File: BLL/Services/DirectoryService.cs [hunk 4]
```diff
@@ -88,19 +101,41 @@ public CityDTO GetCityById(int cityId)
...
             return _mapper.Map<List<City>, List<CityDTO>>(_db.Cities.GetAll().ToList());
         }
 
-        //public StatusDTO GetStatusById(int statusId)
-        //{
-        //    return _mapper.Map<Status, StatusDTO>(_db.Statuses.Get(statusId));
-        //}
+        public List<CityDTO> GetAllCitiesByCountryId(int countryId)
+        {
+            var statesInCountry = _db.States.GetAll().Where(x => x.Country_Id == countryId).ToList();
```
Comment (VasiliNesterenko) [inline BLL/Services/DirectoryService.cs:119]: it's very bad approach
the whole method can be replaced with single code line:

return _mapper.Map<List<City>, List<CityDTO>>(_db.Cities.FindWithSpecificationPattern(new CitySpecification()).Where(x => x.State.Country_Id == countryId).ToList());

all you need is just to include State and State.Country in the CitySpecification class

**Review by VasiliNesterenko — CHANGES_REQUESTED** [review]
it needs to be refactored

File: BLL/DTO/CandidateDTO.cs (hunk 1 above)
Comment (Alexy-Def) [reply BLL/DTO/CandidateDTO.cs:21]: We anyway need this change :(

**Review by VasiliNesterenko — APPROVED** [review]

**Review by AnnKirienko — APPROVED** [review]

**Pull request merged**